daml/Marketplace/ExchangeParticipant.daml

```diff
@@ -160,3 +161,58 @@ template ExchangeParticipant
           orderRequestCid <- create OrderRequest with ..
 
           return (orderRequestCid, debitRequestCid)
+
+      nonconsuming ExchangeParticipant_PlaceClearedBid : ContractId ClearedOrderRequest
+        with
+          ccp : Party
+          pair : IdPair
+          amount : Decimal
+          price : Decimal
+        do
+          (_, baseToken) <- fetchByKey @Token pair._1
+          let qty = roundBankers baseToken.quantityPrecision $ amount / price
+
+          let order = ClearedOrder with
+                isBid = True, qty, status = "New", orderId = -1, ..
+
+          exerciseByKey @CCPCustomer (ccp, operator, exchParticipant) CCPCustomer_MakeClearedOrder
+            with pair, price, order
+
+      nonconsuming ExchangeParticipant_PlaceClearedOffer : ContractId ClearedOrderRequest
+        with
+          ccp : Party
+          pair : IdPair
+          amount : Decimal
+          price : Decimal
+        do
+          (_, baseToken) <- fetchByKey @Token pair._1
+          let qty = roundBankers baseToken.quantityPrecision amount
+
+          let order = ClearedOrder with
+                isBid = False, qty, status = "New", orderId = -1, ..
+
+          exerciseByKey @CCPCustomer (ccp, operator, exchParticipant) CCPCustomer_MakeClearedOrder
+            with pair, price, order
```

Comment:
I don't know if this is "elegant" or "hacky", but I made sure that an investor placing an order in a cleared market has the authorization of the CCP by moving the `MakeClearedOrder` choice to its `CCPCustomer` contract. 

if the investor is not a customer, they won't have such a contract, and the trade placement is blocked by the key not found error!

Reply:
This works, you could also use `lookupByKey` which returns an `Optional` to have a clearer abort message (and maybe keep it in the `ExchangeParticipant` function):

```
optCCPCustomer <- lookupByKey @CCPCustomer etc. etc.
assertMsg "Participant must be a customer etc etc" $ isSome optCCPCustomer
```

Reply:
For some reason, `lookupByKey` wasn't working even when the contract did exist. I forgot what the error was exactly (tried it a week ago), but some detail about disclosure and contract privacy I think. 

I put some frontend validation in place, so if this situation happens we stop it before exercising, and we still avoid a random DAML LF error dump on the screen.